doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] conf the PHY configuration
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*config_phy)(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf);
+
+    /**
+     * @brief Set the transceiver PHY state
+     *
+     * @pre the device is not sleeping
+     *
+     * @note the implementation MUST block until the state change occurs.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] state the new state
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_trx_state)(ieee802154_dev_t *dev, ieee802154_trx_state_t state);
+
+```
+
+### Device State Management
+```c
+    /**
+     * @brief Set the sleep state of the device (sleep or awake)
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] sleep whether the device should sleep or not.
+     *
+     * @post the transceiver state is @ref IEEE802154_TRX_STATE_TRX_OFF,
+     *       regardless of the value of @p sleep
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_sleep)(ieee802154_dev_t *dev, bool sleep);
+
+    /**
+     * @brief Process radio IRQ.
+     *
+     * This function calls the @ref ieee802154_cb_t::cb function with
+     * the corresponding event type.
+     *
+     * @note if the device is sleeping, this function should do nothing
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     */
+    void (*irq_handler)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Start the device
+     *
+     * @pre the init function of the radio succeeded.
+     *
+     * This function puts the radio in a state where it can be operated. It
+     * should enable interrupts and set the transceiver state to
+     * @ref IEEE802154_TRX_STATE_TRX_OFF
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*start)(ieee802154_dev_t *dev);
+```
+
+### Caps and Optional Functions
+```c
+    /**
+     * @brief Get a cap from the radio
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param cap cap to be checked
+     *
+     * @return true if the radio supports the cap
+     * @return false otherwise
+     */
+    bool (*get_cap)(ieee802154_dev_t *dev, ieee802154_rf_caps_t cap);
+
+    /**
+     * @brief Set IEEE802.15.4 promiscuous mode
+     *
+     * @pre the device is not sleeping
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       hardware address filtering.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] enable whether the promiscuous mode should be enabled or not.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_promiscuous)(ieee802154_dev_t *dev, bool enable);
```

Comment:
Can we have this as first element? It would be super nice if the struct of the IEEE 802.15.4 HAL implementation would overlap with a Wireless Network Device HAL. Otherwise (assuming the idea of "object orientation" is agreed upon), one might end up having to device two structs.

This feature is certainly something I would put into the generic Wireless Network Device HAL.

Reply:
Yes, we can rearrange it as we want here